optionally prevent cross-document attention

sohamparikh opened the pull request:
# ✨ Description

Closes #62

Optionally prevent cross-document attention in packed sequences.
Introduce two boolean options in config:
1.  `data.sampling.per_document_positions`: Creates `position_ids` local to each document, instead of global `position_ids` assumed for the whole sequence. Document boundaries are determined by `tokenizer.bod_id` which requires specifying `data.tokenizer` as well
2.  `model.base_model.transformer.prevent_cross_document_attention`: Applies attention mask to each document separately using [flash_attn_varlen_func](https://github.com/Dao-AILab/flash-attention/blob/d82bbf26924c492064af8b27ab299ff4808d1bf6/flash_attn/flash_attn_interface.py#L1375).

~**Note**: This does not currently work with `sequence_data_parallel > 1`. `flash_attn_varlen_func` assumes corresponding indices in both `cu_seqlens_q` and `cu_seqlens_k` refer to the same sequence. This is not currently handled for sequence data parallelism~
Works with `sequence_data_parallel>1`

wandb runs on SmolLM-135M:
Cross-document attention (before): https://wandb.ai/akshaykalkunte/slam/runs/v6x39cllhe2xgpd1
Per-document attention (after): https://wandb.ai/akshaykalkunte/slam/runs/xpfv9w3t2gtt4z4x
![image](https://github.com/user-attachments/assets/2c7c8469-e6bb-4984-a133-b4a45d616d57)
![image](https://github.com/user-attachments/assets/8dfb28fd-6cfe-4265-8002-07df3c34b312)

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

jlamypoirier commented on fast_llm/models/gpt/model.py:
```diff
@@ -228,7 +234,26 @@ def preprocess(
                 # TODO: Avoid multiple contiguous calls?
                 tokens = batch.token_ids[:, sequence_k - sequence_q : sequence_k].contiguous()
-
+            if batch.position_ids is not None:
```
This is only active with `sequence_first`, so you should ensure `sequence_first` is `True` for varlen . See `preprocessed_meta`

sohamparikh replied:
I'm confused, this is executed regardless of `sequence_first` currently. varlen shouldn't depend on it either

jlamypoirier replied:
Ah I misread the diff, nevermind

sohamparikh commented on fast_llm/data/dataset/gpt/sampled.py:
```diff
@@ -491,7 +498,14 @@ def __getitem__(self, idx: int) -> typing.Any:
             spans = None
-        return GPTSample(token_ids=token_ids, loss_masking_spans=spans)
+        if self._config.per_document_positions:
+            start_indices = np.where(token_ids == self._tokenizer.bod_id)[0][:-1]
```
I changed it to using the bos token since we wanted to also make it work with sequences packed and padded outside of fast-llm, as a quick and dirty hack. Although I think that's not a priority anymore, since I'm also working on a padding PR this week.

@tscholak do you think there is still a reason to do this or can we just wait for padding to be available in fast-llm as well?

tscholak replied:
Thanks @sohamparikh, let's wait for padding :)

jlamypoirier commented on fast_llm/layers/language_model/embedding.py:
```diff
@@ -76,7 +75,6 @@ def __init__(
 
     @torch.compile
     def _forward(self, input_: torch.Tensor, position_ids: torch.Tensor | None) -> torch.Tensor:
-        Assert.eq(position_ids is not None, self._use_absolute_position_embeddings)
```
Preferably yes, this is for extra safety

sohamparikh commented on fast_llm/data/dataset/gpt/sampled.py:
```diff
@@ -337,9 +338,14 @@ def __getitem__(self, index: int) -> typing.Any:
         loss_masking_spans = (
             np.stack(loss_masking_spans, dtype=np.int32) if self._config.use_loss_masking_spans else None
         )
+        position_ids = (
```
i did it since `position_ids` are more commonly understood. Makes sense to use seqlens instead

sohamparikh commented:
@jlamypoirier I've addressed your comments:
1. Simpler config: `batch.variable_sequence_lengths` is the only option
2. Use `seqlens` instead of `position_ids` in `GPTSample` and `GPTBatch`. `position_ids` are only created for absolute position embeddings
3. `cu_seqlens_*` and `max_seqlens_*` are now created inside `FlashAttnVarlenPreprocessor`. This is to mainly avoid `preprocess` getting too big and messy
4. Removed flash_attn wrapper
5. Hacked a workaround for `sequence_data_parallel > 1` with varlen (see loss curves below). It's not super efficient, but should work okay for lower parallelism


![image](https://github.com/user-attachments/assets/40217ce9-b20a-481a-bd0d-5bdc1d76c80c)

jlamypoirier commented on fast_llm/layers/transformer/attention.py:
```diff
@@ -340,16 +343,57 @@ def forward(self, input_: torch.Tensor, kwargs: dict[str, typing.Any]) -> tuple[
+                if (cu_seqlens_q := kwargs.get(TransformerKwargs.cu_seqlens_q, None)) is not None:
+                    out_dims = query.size()
+                    query = query.view(-1, query.size(-2), query.size(-1))
+                    if key.size(-3) != out_dims[-3]:
```
I'm not following, what is this for exaclty?

sohamparikh replied:
it's a check for `sequence_data_parallelism`. If number of key/query tokens are the same, we want to avoid the extra computation in this block

sohamparikh replied:
For some more context, the varlen function expects the same number of sequences for key/queries, so this block slices kv pairs accordingly (instead of including all past kv pairs)

jlamypoirier replied:
Can't you just set `seqlen_q=0` for the sequences outside the current range?

sohamparikh replied:
yes, that's much simpler. Thanks!

jlamypoirier commented on fast_llm/layers/transformer/preprocessing.py:
```diff
@@ -254,3 +260,73 @@ def preprocess_meta(self, kwargs: dict[str, typing.Any]) -> None:
+        self._distributed_config = self._tensor_space.distributed_config
+        assert self._config.do_use_flash_attention(self._distributed_config)
+
+    def preprocess(self, kwargs: dict[str, typing.Any]) -> None:
```
That's a big one, consider adding comments and maybe a test or two?

sohamparikh replied:
done

jlamypoirier reviewed (commented):
Looks good, got a few comments on the details

sohamparikh commented on fast_llm/data/data/gpt/config.py:
```diff
@@ -56,11 +56,7 @@ class GPTDataConfig(DataConfig, GPTLegacyConfig):
-        desc="Read and use loss masking spans from the dataset, if present.",
-        hint=FieldHint.feature,
-    )
+    variable_sequence_lengths: bool | None = Field(default=None, desc="Use ")
```
this is unused, `GPTSamplingData.variable_sequence_lengths` is used in `GPTSampledIndexedDataset` and `LegacyGPTSampledIndexedDataset`

jlamypoirier commented on fast_llm/data/dataset/gpt/sampled.py:
```diff
@@ -491,7 +500,15 @@ def __getitem__(self, idx: int) -> typing.Any:
-        return GPTSample(token_ids=token_ids, loss_masking_spans=spans)
+        seqlens = (
+            np.array(
+                [sample.token_ids.size - (idx == len(sample_list) - 1) for idx, sample in enumerate(sample_list)],
```
I see, thanks!

sohamparikh commented on fast_llm/data/dataset/gpt/config.py:
```diff
@@ -62,6 +62,11 @@ class GPTSamplingConfig(SamplingConfig):
         desc="Shuffling strategy.",
         hint=FieldHint.feature,
     )
+    per_document_positions: bool | None = Field(
```
Now that we have a single flag, does it make sense to revert to something more indicative like `prevent_cross_document_attention`, or `document_local_attention`? Because we're still packing variable length sequences even with `variable_sequence_length=False`

@tscholak do you have any recommendations?

jlamypoirier reviewed (changes requested):
~I don't think your approach for sequence-data-parallel makes sense. It simply splits into independent sequences, so it's just plain data parallelism. It's also very inefficient, and only works when there are lots of small sequences, in which case sequence-data-parallel is useless to begin with.~

~What you need to do instead is adjust `cu_seqlens_k` to the correct values, `_flash_attn_varlen_func` is made for this and will handle the rest~

Edit: I think I got your implementation wrong, looking again

jlamypoirier commented on fast_llm/data/data/gpt/data.py:
```diff
@@ -58,6 +66,7 @@ def __init__(
         distributed_config: DistributedConfig,
         vocab_size: int,
         max_sequence_length: int,
+        use_document_boundaries: bool | None = None,
```
Not convinced about this new name. How about `separate_documents`? Or `separate_sequences`, to make it non-specific to language models?

jlamypoirier commented on fast_llm/layers/transformer/config.py:
```diff
@@ -473,6 +478,11 @@ class TransformerConfig(TransformerArchitectureConfig, BaseModelConfig):
     use_flash_attention: bool = Field(
         default=True, desc="Enable Flash Attention if possible.", hint=FieldHint.optional
     )
+    isolate_document_context: bool = Field(
```
I think it was better in the BatchConfig, since it's only used for preprocessing and not by the transformer directly. Also same comment about naming.

tscholak replied:
Let's call this `enable_cross_document_attention`, flip the boolean and make the default `True`.

jlamypoirier replied:
Guess that works. Or simply `cross_document_attention`

jlamypoirier reviewed (approved):
Some minor comments, otherwise LGTM. Great work!
